### PR TITLE
8292596: Make SymbolHashMap a ResourceHashtable

### PR DESCRIPTION
**Pull request by coleenp**
Please review this straightforward change to use ResourceHashtable for SymbolHashMap, which is used by the Class File Reconstituter.  The table maps Symbol* to constant pool indices for loaded classes (Symbol refcounting unnecessary as the class is not unloaded because it's in use).
Tested with jvmti tests locally and tier 1-3.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292596](https://bugs.openjdk.org/browse/JDK-8292596): Make SymbolHashMap a ResourceHashtable

### Reviewers
 * [Harold Seigel](https://openjdk.org/census#hseigel) (@hseigel - **Reviewer**) ⚠️ Review applies to [c0f966bb](https://git.openjdk.org/jdk/pull/9917/files/c0f966bbb5db6f1427e44f906cc89115a8b47524)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [c0f966bb](https://git.openjdk.org/jdk/pull/9917/files/c0f966bbb5db6f1427e44f906cc89115a8b47524)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9917/head:pull/9917` \
`$ git checkout pull/9917`

Update a local copy of the PR: \
`$ git checkout pull/9917` \
`$ git pull https://git.openjdk.org/jdk pull/9917/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9917`

View PR using the GUI difftool: \
`$ git pr show -t 9917`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9917.diff">https://git.openjdk.org/jdk/pull/9917.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/oops/constantPool.hpp [hunk 1]
```diff
@@ -914,88 +914,22 @@ class ConstantPool : public Metadata {
...
   // Default number of entries in the table
-  enum SymbolHashMap_Constants {
-    _Def_HashMap_Size = 256
-  };
-
-  int                   _table_size;
-  SymbolHashMapBucket*  _buckets;
-
-  void initialize_table(int table_size);
+  ResourceHashtable<const Symbol*, u2, 256, ResourceObj::C_HEAP, mtSymbol, Symbol::compute_hash> _table;
```
Comment (hseigel) [inline src/hotspot/share/oops/constantPool.hpp:921]: The comment at new line 920, about 'number of entries' is no longer is accurate.

Comment (coleenp) [reply src/hotspot/share/oops/constantPool.hpp:921]: Thank you for catching that.

**Review by hseigel — APPROVED** [review]
Changes look good!
Thanks, Harold

File: src/hotspot/share/oops/constantPool.hpp [hunk 2]
```diff
@@ -914,88 +914,19 @@ class ConstantPool : public Metadata {
...
-  static unsigned int compute_hash(const char* str, int len) {
-    unsigned int hash = 0;
-    while (len-- > 0) {
-      hash = 31*hash + (unsigned) *str;
-      str++;
-    }
-    return hash;
+  void add_entry(const Symbol* sym, u2 value) {
+    bool created;
+    _table.put_if_absent(sym, value, &created);
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/constantPool.hpp:923]: Pre-existing: Just an observation but it is very poor API naming to have `add_entry` when the semantics are actually `add_if_absent`.

**Review by dholmes-ora — APPROVED** [review]
These simplifications looks great.

A couple of drive-by- comments, one of which is we really should decide on table versus map in the naming here.

Thanks.

**Comment by hseigel** [comment]
Still looks good!
Thanks, Harold

**Comment by coleenp** [comment]
Thanks Harold and David!
/integrate

**Pull request closed without merging**